pkg/operator/controllers/checkers/serviceprincipalchecker/controller_test.go

```diff
@@ -0,0 +1,132 @@
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"errors"
+	"reflect"
+	"testing"
+	"time"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	arofake "github.com/Azure/ARO-RP/pkg/operator/clientset/versioned/fake"
+	checkercommon "github.com/Azure/ARO-RP/pkg/operator/controllers/checkers/common"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient"
+	"github.com/Azure/ARO-RP/pkg/util/cmp"
+	utillog "github.com/Azure/ARO-RP/pkg/util/log"
+)
+
+type fakeChecker func(ctx context.Context, AZEnvironment string) error
+
+func (fc fakeChecker) Check(ctx context.Context, AZEnvironment string) error {
+	return fc(ctx, AZEnvironment)
+}
+
+func TestReconcile(t *testing.T) {
+	ctx := context.Background()
+
+	tests := []struct {
+		name                 string
+		controllerDisabled   bool
+		checkerReturnErr     error
+		wantConditionStatus  operatorv1.ConditionStatus
+		wantConditionMessage string
+		wantErr              string
+		wantResult           reconcile.Result
+	}{
+		{
+			name:                 "no errors",
+			wantConditionStatus:  operatorv1.ConditionTrue,
+			wantConditionMessage: "service principal is valid",
+			wantResult:           reconcile.Result{RequeueAfter: time.Hour},
+		},
+		{
+			name:                 "check failed with an error",
+			wantConditionStatus:  operatorv1.ConditionFalse,
+			wantConditionMessage: "fake basic error",
+			checkerReturnErr:     errors.New("fake basic error"),
+			wantErr:              "fake basic error",
+			wantResult:           reconcile.Result{RequeueAfter: time.Hour},
+		},
+		{
+			name:                "controller disabled",
+			controllerDisabled:  true,
+			wantConditionStatus: operatorv1.ConditionUnknown,
+			wantResult:          reconcile.Result{},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			instance := &arov1alpha1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: arov1alpha1.SingletonClusterName,
+				},
+				Spec: arov1alpha1.ClusterSpec{
+					AZEnvironment: azureclient.PublicCloud.Environment.Name,
+					OperatorFlags: arov1alpha1.OperatorFlags{
+						checkercommon.ControllerEnabled: "true",
+					},
+				},
+			}
+			if tt.controllerDisabled {
+				instance.Spec.OperatorFlags[checkercommon.ControllerEnabled] = "false"
+			}
+
+			arocli := arofake.NewSimpleClientset(instance)
+
+			r := &Reconciler{
+				log:  utillog.GetLogger(),
+				role: "master",
+				checker: fakeChecker(func(ctx context.Context, AZEnvironment string) error {
+					if !reflect.DeepEqual(AZEnvironment, azureclient.PublicCloud.Environment.Name) {
+						t.Error(cmp.Diff(AZEnvironment, azureclient.PublicCloud.Environment.Name))
+					}
+
+					return tt.checkerReturnErr
+				}),
```

Comment:
I am having a hard time following that part, do we need to type cast it as a `fakechecker` ? wouldn't the function satisfy the type directly? (I think it should be an interface if we have to mock it though, as per previous comment)

Reply:
@facchettos I'll answer two related comments here.

> wouldn't the function satisfy the type directly?

No - `checker` field in the `Reconciler` expects an instance which implements `servicePrincipalChecker` interface. The anonymous function on these lines won't statisfy the interface.

Hence we need type casting to `fakeChecker` which does statisfy it.

> (I think it should be an interface if we have to mock it though, as per previous comment)

It already is an interface and we mock it with `fakeChecker`. We do not use `mockgen` from `go-mock` becaue it is unexported interface and only is required once. So I see no point of generating a mock for it.

No about [the above comment](https://github.com/Azure/ARO-RP/pull/2588/files#r1048210367):

> this pattern seems a little weird tbh. what does it achieve, is it a function on a function? not 100% sure I understand it (maybe just missing something)
shouldn't the checker be an interface if we need to mock it to test the reconciler?

This is the same pattern as with `http.HandlerFunc` from the standart library ([source](https://cs.opensource.google/go/go/+/refs/tags/go1.19.4:src/net/http/server.go;l=2101-2110)).

We just make a function to statisfy an interface with one method.

Reply:
is it really simplifying if we compare it to that though?
```go
type mockChecker struct {
  returnError error
}
func (fc mockChecker) Check(ctx context.Context, AZEnvironment string) error {
	if !reflect.DeepEqual(AZEnvironment, azureclient.PublicCloud.Environment.Name) {
		t.Error(cmp.Diff(AZEnvironment, azureclient.PublicCloud.Environment.Name))
	}

	return returnError
}

//and in the test
fakeChecker := mockChecker{tt.checkReturnError}
r := &Reconciler{
				log:  utillog.GetLogger(),
				role: "master",
				checker: fakeChecker,
}
```

Reply:
@facchettos I think with this approach `mockChecker` would have to be defined within nested `t.Run` somehow. Or you would have to store `t *testing.T` inside of `mockChecker` struct. Otherwise `t.Error` will be giving you wrong context in case of failure.

So instead of defining the whole type inside `t.Run` I just define an anonymous function and use `http.HandlerFunc` pattern.

I think what we have right now in this PR is marginally better than struct approach.